soccer/planning/planner/CollectPlanner.cpp

```diff
@@ -0,0 +1,139 @@
+    _distCutoffToControl =
+        new ConfigDouble(cfg, "Capture/Collect/distCutoffToControl");
+}
+
+using namespace Geometry2d;
+Trajectory CollectPlanner::plan(PlanRequest&& request) {
+    processStateTransitions(request);
+    switch (_collectStates[request.shellID]) {
+        case CollectState::Course:
+            return course(std::move(request));
+        case CollectState::Fine:
+            return fine(std::move(request));
+        case CollectState::Control:
+            return control(std::move(request));
+    }
+    debugThrow("Error: CollectPlanner state machine failed.");
+    return Trajectory{{}};
+}
+Trajectory CollectPlanner::course(PlanRequest&& request) {
+    RobotInstant startInstant = request.start;
+    RotationConstraints rotationConstraints = request.constraints.rot;
+    const Ball& ball = request.context->state.ball;
+    Point approachDir = findApproachDirection(request);
+    Point noisyTargetPoint =
+        ball.pos - approachDir * (Ball_Radius + Robot_MouthRadius +
+                                  *_bufferDistBeforeContact);
+    Point noisyTargetVel =
+        approachDir * (ball.vel.dot(approachDir) + *_touchDeltaSpeed);
+    auto& targetInstant = _courseTargetInstants[request.shellID];
+    if (!targetInstant || targetInstant->pose.position().distTo(
+                              noisyTargetPoint) > *_targetChangeThreshold) {
+        targetInstant = RobotInstant{Pose{noisyTargetPoint, 0},
+                                     Twist{noisyTargetVel, 0}, RJ::Time{0s}};
+    }
+    request.motionCommand = PathTargetCommand{*targetInstant};
+    request.static_obstacles.add(
+        std::make_shared<Circle>(ball.pos, Ball_Radius));
+    Trajectory path = _pathTargetPlanner.plan(std::move(request));
+    PlanAngles(path, startInstant, AngleFns::faceAngle(approachDir.angle()),
+               rotationConstraints);
+    return std::move(path);
+}
+Trajectory CollectPlanner::fine(PlanRequest&& request) {
+    Point approachDir = findApproachDirection(request);
+    const Ball& ball = request.context->state.ball;
+    RobotInstant startInstant = request.start;
+    RotationConstraints rotationConstraints = request.constraints.rot;
+    Point targetPoint =
+        ball.pos - approachDir * (Ball_Radius + Robot_MouthRadius);
+    request.constraints.mot.maxSpeed =
+        ball.vel.dot(approachDir) + *_touchDeltaSpeed;
+    Point targetVel = request.constraints.mot.maxSpeed * approachDir;
+    RobotInstant targetInstant{Pose{targetPoint, 0}, Twist{targetVel, 0},
+                               RJ::Time{0s}};
+    request.motionCommand = PathTargetCommand{targetInstant};
+    Trajectory path = _pathTargetPlanner.plan(std::move(request));
+    PlanAngles(path, startInstant, AngleFns::faceAngle(approachDir.angle()),
+               rotationConstraints);
+    return std::move(path);
+}
+Trajectory CollectPlanner::control(PlanRequest&& request) {
+    Point approachDir = findApproachDirection(request);
```

Comment:
If there is any increase in ball velocity due to touching it, you may get runaway ball behavior because you try to match it's speed. I had to make it only plan once to get around this.